src/metabase/types.clj

```diff
@@ -5,6 +5,7 @@
 (derive :type/Dictionary :type/Collection)
 (derive :type/Array :type/Collection)
 
+(derive :type/Nil :type/*)
```

Comment:
You could just dispatch off of `nil` and hand-off to the same function that handles `:type/*`. That way we don't need to define a `:type/Nil`. Again, it's a one-off type used in a single place that I think is going to confuse a lot of people like potential contributors when they come in and look at `types.clj` and see it.

`:type/Number` is different than something just being an instance of a subclass of `Number` because it can (at least theoretically) be used to mark things that are effectively numbers but encoded as something else, for example text. For example a field might have a base type of `:type/Text` and a special type of `:type/Number`.